distributed/executor.py

```diff
@@ -0,0 +1,140 @@
+from __future__ import print_function, division, absolute_import
+
+from dask.base import tokenize
+from tornado import gen
+from tornado.gen import Return
+from tornado.locks import Event
+from tornado.concurrent import Future
+from tornado.ioloop import IOLoop
+from tornado.iostream import StreamClosedError
+from tornado.queues import Queue
+
+from .core import read, write, connect, rpc, coerce_to_rpc
+from .client import WrappedKey, _gather
+from .dask import scheduler, worker, delete
+from .utils import All
+
+
+log = print
+
+
+class Future(WrappedKey):
+    """ The result of a remotely running computation """
+    def __init__(self, key, event, center):
+        self.key = key
+        self.event = event
+        self.center = center
+        self.status = None
+        self.event = Event()
+
+    def _set_ready(self, status):
+        if status:
+            self.event.set()
+        else:
+            self.event.clear()
+
+    def done(self):
+        return self.event.is_set()
+
+    @gen.coroutine
+    def _result(self):
+        yield self.event.wait()
+        result = yield _gather(self.center, [self.key])
+        raise gen.Return(result[0])
+
+
+class Executor(object):
+    """ Distributed executor with data dependencies
+
+    This executor resembles executors in concurrent.futures but also allows
+    Futures within submit/map calls.
+
+    Provide center address on initialization
+
+    >>> executor = Executor(('127.0.0.1', 8787))
+
+    Use ``submit`` method like normal
+
+    >>> a = executor.submit(add, 1, 2)
+    >>> b = executor.submit(add, 10, 20)
+
+    Additionally, provide results of submit calls (futures) to further submit
+    calls:
+
+    >>> c = executor.submit(add, a, b)
+
+    This allows for the dynamic creation of complex dependencies.
```

Comment:
@eriknw this docstring and the code for `def submit(func, *args, **kwargs)` may help to motivate https://github.com/blaze/dask/issues/781